server/src/test/java/org/elasticsearch/common/util/concurrent/EsExecutorsTests.java

```diff
@@ -859,46 +856,68 @@ public void testScalingWithEmptyCoreAndKeepAliveAndWorkerPoolProbing() {
     }
 
     private void testScalingWithEmptyCore(EsThreadPoolExecutor executor) {
-        try {
-            class Task extends AbstractRunnable {
-                private int remaining;
-                private final CountDownLatch doneLatch;
-
-                Task(int iterations, CountDownLatch doneLatch) {
-                    this.remaining = iterations;
-                    this.doneLatch = doneLatch;
-                }
 
-                @Override
-                public void onFailure(Exception e) {
-                    fail(e);
-                }
+        class TaskScheduler extends SubscribableListener<Void> {
+            final ExecutorService scheduler;
+            final CyclicBarrier cyclicBarrier;
+            final Semaphore taskCompletions;
+            volatile int remaining;
+
+            TaskScheduler(ExecutorService scheduler, int iterations) {
+                this.scheduler = scheduler;
+                this.taskCompletions = new Semaphore(0);
+                this.cyclicBarrier = new CyclicBarrier(executor.getMaximumPoolSize(), () -> remaining--);
+                this.remaining = iterations;
+            }
 
-                @Override
-                protected void doRun() {
-                    if (--remaining == 0) {
-                        doneLatch.countDown();
-                    } else {
-                        logger.trace("--> remaining [{}]", remaining);
-                        final long keepAliveNanos = executor.getKeepAliveTime(TimeUnit.NANOSECONDS);
-                        new Thread(() -> {
-                            if (keepAliveNanos > 0) {
-                                final var targetNanoTime = System.nanoTime() + keepAliveNanos + between(-10_000, 10_000);
-                                while (System.nanoTime() < targetNanoTime) {
-                                    Thread.yield();
+            public void start() {
+                final var keepAliveNanos = executor.getKeepAliveTime(TimeUnit.NANOSECONDS);
+                final var tasks = executor.getMaximumPoolSize();
+                Runnable runnable = () -> {
+                    try {
+                        while (remaining > 0) {
+                            // wait for all scheduler threads to be ready for the next attempt
+                            var first = cyclicBarrier.await(2, TimeUnit.SECONDS) == 0;
+                            executor.execute(taskCompletions::release);
+                            if (first) {
+                                // let the first scheduler (by arrival on the barrier) wait for completion of all tasks
+                                if (taskCompletions.tryAcquire(tasks, 1, TimeUnit.SECONDS) == false) {
```

Comment:
Timeouts shorter than `SAFE_AWAIT_TIMEOUT` (i.e. 10s) are generally shaky in tests because you might just be running on a CPU-starved worker. Could we use 10s here?